083-stretch-cluster.md

```diff
@@ -0,0 +1,155 @@
+
+An example of the KafkaNodePool resource with the new fields might look like:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: controller
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  target:
+    clusterUrl: <K8S Cluster URL>
+    secret: <SecretName>
+  listenerConfig:
+    - configuration:
+          bootstrap:
+            alternativeNames:
+              - bootstrap-ext.<kubernetes infra host>
+              - broker-100.<kubernetes infra host>
+              - broker-101.<kubernetes infra host>
+            host: bootstrap-ext.<kubernetes infra host>
+          brokers:
+          - broker: 100
+            host: broker-100.<kubernetes infra host>
+          - broker: 101
+            host: broker-101.<kubernetes infra host>
+          class: nginx
+        name: connectext
+  roles:
+    - controller
+  storage:
+   .........
+```
+
+#### Kafka changes
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+  annotations:
+    strimzi.io/node-pools: enabled
+    strimzi.io/kraft: enabled
+    strimzi.io/stretch-mode: enabled
+spec:
+  kafka:
+    version: 3.7.0
+    metadataVersion: 3.7-IV4
+    listeners:
+      - port: 9093
+        tls: true
+        name: connectext
+        type: ingress
+```
+
+A new annotation (`stretch-mode: enabled`) will be introduced in Kafka custom resource to indicate when it is representing a stretch Kafka cluster. This approach is similar to how Strimzi currently enables features like KafkaNodePool (KNP) and KRaft mode.
+
+In a stretch Kafka cluster, we'll need bootstrap and broker services to be present on each Kubernetes cluster and be accessible from other clusters. The Kafka reconciler will identify all target clusters from KafkaNodePool resources and create these services in target Kubernetes clusters. This will ensure that even if the central cluster experiences an outage, external clients can still connect to the stretch cluster and continue their operations without interruption.
+
+#### Cross-cluster communication
+Kafka controllers/brokers are distributed across multiple Kubernetes environments and will need to communicate with each other. Currently, the Strimzi Kafka operator defines Kafka listeners for internal communication (controlplane and replication) between brokers/controllers (Kubernetes services using ports 9090 and 9091). The user is not able to influence how these services are set up and exposed outside the cluster. We would remove this limitation and allow users to define how these internal listeners are configured in the Kafka resource, just like they do for Kafka client listeners.
```

Comment:
The thinking was for the Kafka reconciler will detect this is a stretch cluster, and if so, then relax the restrictions that we have in place that limit the minimum listener port to 9090 (instead of 9092). So user can then define listeners for 9090 and 9091 too.

Reply:
I think we still need to be in control of those. The things such as security etc. should remain under our control. We will of course need some additional configs to define the networking used for the stretch cluster etc. But I do not think it is as simple as freeing the port numbers.